### PR TITLE
[ML] Inference configs for NLP models

### PR DESCRIPTION
**Pull request by dimitris-athanasiou**
Introduce inference configs for NLP models.

When a PyTorch model is put, the config now expects
a different inference config per task type. Thus, we
have a `ner`, `fill_mask`, and `sentiment_analysis`
config. In addition, the tokenization parameters have
been grouped together and are now part of the relevant
inference config objects. Thus the vocabulary can now
be on a document on its own. A new vocabulary config
object allows the user to specify the location of the
vocabulary document.

An example of how the request to put an NLP model looks like follows:

```
{
    "description": "NER model",
    "model_type": "pytorch",
    "inference_config": {
        "ner": {
            "vocabulary": {
                "index": "vocabulary_index",
                "id": "vocabulary_doc_id"
            }
        }
    },
    "location": {
        "index": {
            "name": "ner"
        }
    }
}
```

Finally, instead of a model config file, we now just need the vocabulary whose
should just be:

```
{
  "vocab": [ ... ]
}
```

### CODE REVIEW

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelConfig.java [hunk 1]
```diff
@@ -181,7 +181,8 @@
         this.description = description;
         this.tags = Collections.unmodifiableList(ExceptionsHelper.requireNonNull(tags, TAGS));
         this.metadata = metadata == null ? null : Collections.unmodifiableMap(metadata);
-        this.input = ExceptionsHelper.requireNonNull(input, INPUT);
+        this.input = (input == null && modelType == TrainedModelType.PYTORCH) ?
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelConfig.java:184]: ++ A user should not have to think about which possible field will be used for input when defining the model config, it is easier to define at the point of use (e.g. in an ingest processor) when the field names are known.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfig.java [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Objects;
+
+public class NerConfig implements NlpConfig {
+
+    public static final String NAME = "ner";
+
+    private static final ParseField VOCABULARY = new ParseField("vocabulary");
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfig.java:28]: These parse fields are common to the different configs. Can they be moved to `NlpConfig` and shared?

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfig.java [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+        tokenizationParams = new TokenizationParams(in);
+        classificationLabels = in.readOptionalStringList();
+    }
+
+    @Override
+    public XContentBuilder toXContent(XContentBuilder builder, Params params) throws IOException {
+        builder.startObject();
+        builder.field(VOCABULARY.getPreferredName(), vocabularyConfig);
+        builder.field(TOKENIZATION_PARAMS.getPreferredName(), tokenizationParams);
+        if (classificationLabels != null) {
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfig.java:77]:
```suggestion
        if (classificationLabels != null && classificationLabels.isEmpty() == false) {
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TokenizationParams.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+    private final boolean withSpecialTokens;
+    private final int maxSequenceLength;
+
+    public static TokenizationParams createDefault() {
+        return new TokenizationParams(null, null, null);
+    }
+
+    public TokenizationParams(@Nullable Boolean doLowerCase, @Nullable Boolean withSpecialTokens, @Nullable Integer maxSequenceLength) {
+        if (maxSequenceLength != null && maxSequenceLength <= 0) {
+            throw new IllegalArgumentException("[{}] must be positive");
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TokenizationParams.java:49]:
```suggestion
            throw new IllegalArgumentException("[max_sequence_length] must be positive");
```

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java [hunk 5]
```diff
@@ -106,57 +109,66 @@ private void doStartDeployment(TrainedModelDeploymentTask task, ActionListener<T
...
+                    },
+                    listener::onFailure
+                ));
             },
             listener::onFailure
         );
 
-        SearchRequest searchRequest = taskConfigSearchRequest(taskConfigDocId, task.getIndex());
-        executeAsyncWithOrigin(client, ML_ORIGIN, SearchAction.INSTANCE, searchRequest, configListener);
+        executeAsyncWithOrigin(client, ML_ORIGIN, GetTrainedModelsAction.INSTANCE, new GetTrainedModelsAction.Request(task.getModelId()),
```
Comment (davidkyle) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java:144]: much better to be doing it like this.

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java [hunk 6]
```diff
@@ -157,12 +155,10 @@ public static TaskParams fromXContent(XContentParser parser) {
         private static final ByteSizeValue MEMORY_OVERHEAD = ByteSizeValue.ofMb(270);
 
         private final String modelId;
-        private final String index;
         private final long modelBytes;
 
-        public TaskParams(String modelId, String index, long modelBytes) {
+        public TaskParams(String modelId, long modelBytes) {
```
Comment (davidkyle) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/action/StartTrainedModelDeploymentAction.java:160]: Nice to remove the index from the task params it was a bit of an oddity

**Review by davidkyle — APPROVED** [review]
LGTM

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelConfig.java (hunk 1 above)
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/TrainedModelConfig.java:184]: If this is going to be our design, we should have the input default based on the model type, and consequently set there. 

```
this.input = ExceptionsHelper.requireNonNull(input == null ? modelType.defaultInput() : input, INPUT);
```

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfig.java [hunk 7]
```diff
@@ -0,0 +1,141 @@
...
+    public void writeTo(StreamOutput out) throws IOException {
+        vocabularyConfig.writeTo(out);
+        tokenizationParams.writeTo(out);
+        out.writeOptionalStringCollection(classificationLabels);
+    }
+
+    @Override
+    public boolean isTargetTypeSupported(TargetType targetType) {
+        return false;
+    }
```
Comment (benwtrent) [inline x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfig.java:94]: This and all other of these checks, I am thinking we need to add a new `TargetType` for `NLP` or something. We shouldn't allow these types of configs for the boosted tree models. 

I am not 100% sure the best way here.

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/SentimentAnalysisConfigTests.java [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    protected SentimentAnalysisConfig mutateInstanceForVersion(SentimentAnalysisConfig instance, Version version) {
+        return instance;
+    }
+
+    public static SentimentAnalysisConfig createRandom() {
+        return new SentimentAnalysisConfig(
+            VocabularyConfigTests.createRandom(),
+            TokenizationParamsTests.createRandom(),
+            randomList(5, () -> randomAlphaOfLength(10))
```
Comment (benwtrent) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/SentimentAnalysisConfigTests.java:51]: This two should be randomly null

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfigTests.java [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+    @Override
+    protected NerConfig mutateInstanceForVersion(NerConfig instance, Version version) {
+        return instance;
+    }
+
+    public static NerConfig createRandom() {
+        return new NerConfig(
+            VocabularyConfigTests.createRandom(),
+            TokenizationParamsTests.createRandom(),
+            randomList(5, () -> randomAlphaOfLength(10))
```
Comment (benwtrent) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfigTests.java:51]: These should be randomly null

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TokenizationParamsTests.java [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        return createRandom();
+    }
+
+    @Override
+    protected TokenizationParams mutateInstanceForVersion(TokenizationParams instance, Version version) {
+        return instance;
+    }
+
+    public static TokenizationParams createRandom() {
+        return new TokenizationParams(randomBoolean(), randomBoolean(), randomIntBetween(1, 1024));
```
Comment (benwtrent) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/TokenizationParamsTests.java:48]: length should be randomly null

File: x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/BertPassThroughConfigTests.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+        return createRandom();
+    }
+
+    @Override
+    protected BertPassThroughConfig mutateInstanceForVersion(BertPassThroughConfig instance, Version version) {
+        return instance;
+    }
+
+    public static BertPassThroughConfig createRandom() {
+        return new BertPassThroughConfig(VocabularyConfigTests.createRandom(), TokenizationParamsTests.createRandom());
```
Comment (benwtrent) [inline x-pack/plugin/core/src/test/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/BertPassThroughConfigTests.java:48]: tokenization should be randomly null

File: x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java [hunk 12]
```diff
@@ -301,8 +311,12 @@ synchronized void stopProcess() {
             };
         }
 
-        void loadModel(ActionListener<Boolean> listener) {
-            process.get().loadModel(modelId, index, stateStreamer, listener);
+        void loadModel(TrainedModelLocation modelLocation, ActionListener<Boolean> listener) {
```
Comment (benwtrent) [inline x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/deployment/DeploymentManager.java:314]: ++

File: x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfig.java (hunk 7 above)
Comment (dimitris-athanasiou) [reply x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ml/inference/trainedmodel/NerConfig.java:94]: Agreed. I'm planning to address this in a follow-up PR where we have validation on the model_type -> inference_config combinations.

**Review by benwtrent — APPROVED** [review]

**Comment by dimitris-athanasiou** [comment]
@elasticmachine update branch

**Comment by dimitris-athanasiou** [comment]
run elasticsearch-ci/part-2

**Pull request merged**